File: mac-native/Sources/DataMoverMac/HistoryStore.swift
import Foundation

/// Istoricul copierilor efectuate — data, numele folderului (Proiect+Card),
/// sursele si destinatiile folosite, plus cate fisiere OK/sarite/esuate.
/// Persistat in Application Support, ca sa ramana intre lansari ale aplicatiei
/// (echivalentul "istoricului" pe care il avea si aplicatia Windows).
struct HistoryEntry: Codable, Identifiable {
    var id: String { "\(dateText)-\(folderName)" }
    let dateText: String
    let folderName: String
    let sourcesSummary: String
    let destSummary: String
    let okCount: Int
    let skipCount: Int
    let failCount: Int
    // Cai complete (2026-08-28) - sourcesSummary/destSummary raman doar
    // pentru afisare scurta; astea sunt necesare ca sa poti deschide
    // direct sursa/destinatia din istoric (Finder). `decodeIfPresent`
    // pastreaza compatibilitatea cu intrari vechi de istoric, salvate
    // inainte de aceasta schimbare (fara aceste campuri).
    var sourcePaths: [String] = []
    var destinationPaths: [String] = []
    /// Radacinile REALE create la fiecare destinatie (destinationPaths[i]
    /// + folderName) - ce ar trebui deschis efectiv in Finder, nu discul
    /// intreg.
    var destinationTargetPaths: [String] = []

    enum CodingKeys: String, CodingKey {
        case dateText, folderName, sourcesSummary, destSummary, okCount, skipCount, failCount
        case sourcePaths, destinationPaths, destinationTargetPaths
    }

    init(dateText: String, folderName: String, sourcesSummary: String, destSummary: String,
         okCount: Int, skipCount: Int, failCount: Int,
         sourcePaths: [String] = [], destinationPaths: [String] = [], destinationTargetPaths: [String] = []) {
        self.dateText = dateText
        self.folderName = folderName
        self.sourcesSummary = sourcesSummary
        self.destSummary = destSummary
        self.okCount = okCount
        self.skipCount = skipCount
        self.failCount = failCount
        self.sourcePaths = sourcePaths
        self.destinationPaths = destinationPaths
        self.destinationTargetPaths = destinationTargetPaths
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        dateText = try c.decode(String.self, forKey: .dateText)
        folderName = try c.decode(String.self, forKey: .folderName)
        sourcesSummary = try c.decode(String.self, forKey: .sourcesSummary)
        destSummary = try c.decode(String.self, forKey: .destSummary)
        okCount = try c.decode(Int.self, forKey: .okCount)
        skipCount = try c.decode(Int.self, forKey: .skipCount)
        failCount = try c.decode(Int.self, forKey: .failCount)
        sourcePaths = try c.decodeIfPresent([String].self, forKey: .sourcePaths) ?? []
        destinationPaths = try c.decodeIfPresent([String].self, forKey: .destinationPaths) ?? []
        destinationTargetPaths = try c.decodeIfPresent([String].self, forKey: .destinationTargetPaths) ?? []
    }
}

final class HistoryStore: ObservableObject {
    static let shared = HistoryStore()

    private let fileURL: URL

    @Published private(set) var entries: [HistoryEntry] = []

    private init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("DataMover", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        fileURL = base.appendingPathComponent("history.json")
        load()
    }

    private func load() {
        guard let data = try? Data(contentsOf: fileURL),
              let decoded = try? JSONDecoder().decode([HistoryEntry].self, from: data) else { return }
        entries = decoded
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    func record(folderName: String, sources: [String], destinations: [String],
                okCount: Int, skipCount: Int, failCount: Int) {
        let df = DateFormatter()
        df.dateFormat = "dd.MM.yyyy HH:mm"
        let entry = HistoryEntry(
            dateText: df.string(from: Date()),
            folderName: folderName,
            sourcesSummary: sources.map { ($0 as NSString).lastPathComponent }.joined(separator: ", "),
            destSummary: destinations.map { ($0 as NSString).lastPathComponent }.joined(separator: ", "),
            okCount: okCount, skipCount: skipCount, failCount: failCount,
            sourcePaths: sources, destinationPaths: destinations,
            destinationTargetPaths: destinations.map { ($0 as NSString).appendingPathComponent(folderName) }
        )
        entries.append(entry)
        if entries.count > 200 { entries.removeFirst(entries.count - 200) }
        save()
    }

    func delete(_ entry: HistoryEntry) {
        entries.removeAll { $0.id == entry.id }
        save()
    }

    func clearAll() {
        entries.removeAll()
        save()
    }
}

File: windows-native/DataMover.Core/Services/HistoryStore.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DataMover.Core.Services;

public sealed class HistoryEntry
{
    public string DateText { get; set; } = "";
    public string FolderName { get; set; } = "";
    public string SourcesSummary { get; set; } = "";
    public string DestSummary { get; set; } = "";
    public int OkCount { get; set; }
    public int SkipCount { get; set; }
    public int FailCount { get; set; }
    // Cai complete - necesare pentru "Deschide sursa/destinatia" din
    // istoric. Optionale (JsonIgnoreCondition.WhenWritingNull nu e nevoie
    // aici - default "" e suficient pentru compatibilitate cu istoric
    // vechi care oricum nu exista inca pe acest client nou).
    public List<string> SourcePaths { get; set; } = new();
    public List<string> DestinationPaths { get; set; } = new();
    public List<string> DestinationTargetPaths { get; set; } = new();
}

/// Istoric persistat in %AppData%\DataMover\history.json - port 1:1 al
/// HistoryStore.swift (Mac) / ~/.datamover_history.json (Python).
public sealed class HistoryStore
{
    private static readonly Lazy<HistoryStore> _shared = new(() => new HistoryStore());
    public static HistoryStore Shared => _shared.Value;

    private readonly string _filePath;
    public List<HistoryEntry> Entries { get; private set; } = new();

    private HistoryStore()
    {
        var baseDir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "DataMover");
        Directory.CreateDirectory(baseDir);
        _filePath = Path.Combine(baseDir, "history.json");
        Load();
    }

    private void Load()
    {
        try
        {
            if (!File.Exists(_filePath)) return;
            var json = File.ReadAllText(_filePath);
            Entries = JsonSerializer.Deserialize<List<HistoryEntry>>(json) ?? new();
        }
        catch { Entries = new(); }
    }

    private void Save()
    {
        try { File.WriteAllText(_filePath, JsonSerializer.Serialize(Entries)); }
        catch { /* best-effort */ }
    }

    public void Record(string folderName, List<string> sources, List<string> destinations,
        int okCount, int skipCount, int failCount)
    {
        var entry = new HistoryEntry
        {
            DateText = DateTime.Now.ToString("dd.MM.yyyy HH:mm"),
            FolderName = folderName,
            SourcesSummary = string.Join(", ", sources.Select(s => Path.GetFileName(s.TrimEnd('/', '\\')))),
            DestSummary = string.Join(", ", destinations.Select(d => Path.GetFileName(d.TrimEnd('/', '\\')))),
            OkCount = okCount, SkipCount = skipCount, FailCount = failCount,
            SourcePaths = sources, DestinationPaths = destinations,
            DestinationTargetPaths = destinations.Select(d => Path.Combine(d, folderName)).ToList(),
        };
        Entries.Add(entry);
        if (Entries.Count > 200) Entries.RemoveRange(0, Entries.Count - 200);
        Save();
    }

    public void Delete(HistoryEntry entry) { Entries.Remove(entry); Save(); }
    public void ClearAll() { Entries.Clear(); Save(); }
}
